src/Mod/Sketcher/Gui/DrawSketchHandlerArcSlot.h

```diff
@@ -0,0 +1,932 @@
+        if (state() == SelectMode::SeekSecond || state() == SelectMode::SeekThird) {
+            setAngleSnapping(true, centerPoint);
+        }
+        else {
+            setAngleSnapping(false);
+        }
+    }
+
+    virtual void onButtonPressed(Base::Vector2d onSketchPos) override
+    {
+        this->updateDataAndDrawToPosition(onSketchPos);
+        if (canGoToNextMode()) {
+            if (state() == SelectMode::SeekThird) {
+                if (arcAngle > 0) {
+                    endAngle = startAngle + arcAngle;
+                    angleReversed = false;
+                }
+                else {
+                    endAngle = startAngle;
+                    startAngle = startAngle + arcAngle;
+                    angleReversed = true;
+                }
+            }
+            this->moveToNextMode();
+        }
+    }
+
+    virtual void createShape(bool onlyeditoutline) override
+    {
+        ShapeGeometry.clear();
+
+        if (radius < Precision::Confusion()) {
+            return;
+        }
+
+        if (state() == SelectMode::SeekSecond) {
+            addCircleToShapeGeometry(toVector3d(centerPoint), radius, isConstructionMode());
+        }
+        else {
+            if (fabs(arcAngle) < Precision::Confusion()) {
+                return;
+            }
+
+            if (state() == SelectMode::SeekFourth && r < Precision::Confusion()) {
+                return;
+            }
+
+            if (constructionMethod() == DrawSketchHandlerArcSlot::ConstructionMethod::ArcSlot) {
+                addArcToShapeGeometry(toVector3d(centerPoint),
+                                      startAngle,
+                                      endAngle,
+                                      radius + r,
+                                      isConstructionMode());
+
+                addArcToShapeGeometry(toVector3d(startPoint),
+                                      angleReversed ? endAngle : startAngle + M_PI,
+                                      angleReversed ? endAngle + M_PI : startAngle + 2 * M_PI,
+                                      r,
+                                      isConstructionMode());
+
+                addArcToShapeGeometry(toVector3d(endPoint),
+                                      angleReversed ? startAngle + M_PI : endAngle,
+                                      angleReversed ? startAngle + 2 * M_PI : M_PI + endAngle,
+                                      r,
+                                      isConstructionMode());
+
+                if (radius - r > Precision::Confusion()) {
+                    addArcToShapeGeometry(toVector3d(centerPoint),
+                                          startAngle,
+                                          endAngle,
+                                          radius - r,
+                                          isConstructionMode());
+                }
+            }
+            else {
+                auto arc1 = std::make_unique<Part::GeomArcOfCircle>();

```

Comment:
Ah! I see the fact that addArcToShapeGeometry does not return a reference is making your code more verbose. I will try to improve that.

Reply:
In:
https://github.com/FreeCAD/FreeCAD/pull/11358 

The helper function is extended to return a reference. So you can call addArcToShapeGeometry and get as returned value a reference to the added arc. This you may use to modify the added arc, or to get endpoints. That way, you do not need to manually create the shape, which is less concise and error prone.

Reply:
Great

Reply:
Have you actually tried this? I cannot get it to work because you return a Part::Geometry unique ptr and I cannot find how to cast a unique ptr.

Reply:
Besides if you look at arc2, I'm adding the arc2 after the 2 lines on purpose because sometimes arc2 doesn't exist. So using addArcToShapeGeometry here makes the code convoluted.

Reply:
static_casting and dynamic_casting of unique_ptr you do via std::static_pointer_cast and std::dynamic_pointer_cast: 

https://en.cppreference.com/w/cpp/memory/shared_ptr/pointer_cast

So in your case, it appears you want something like:

```
auto geo = addArcToShapeGeometry(toVector3d(centerPoint),
                                                      startAngle,
                                                      endAngle,
                                                      r,
                                                      isConstructionMode());

auto arc2 = std::static_pointer_cast<Part::GeomArcOfCircle>(geo);
```

This syntax I can improve into a one liner in a subsequent PR.

I did not look in detail into the order pushing to the vector. If you want to add the arc at the end, then this will not work for you... unless you can add first the lines and then construct the arc afterwards.

If this facility allows you to write better and less cluttered code by reusing framework code (and thus will also be less error prone code), then you should use the facility. If the facility gets in the middle of your flow, then this specific instance may not be the best for this facility.

My intention was not to make your life miserable, but easier. If it is not easier (after understanding the static casting of smart pointers),

Reply:
I will push a new PR, in which you get a type specific naked pointer. This is probably the best solution. So if you ask for a Part::GeomArcOfCircle, it will create the arc, put it inside a smart pointer and add it to the ShapeGeometry array, and return to you a ```Part::GeomArcOfCircle *```, which you do not need to cast to anything. Of course, memory management and ownership is kept with the smart pointer, so you need not take care of any kind of memory management:

https://github.com/FreeCAD/FreeCAD/pull/11362

Reply:
> My intention was not to make your life miserable, but easier. If it is not easier (after understanding the static casting of smart pointers),

Yes I know that and it does help :)